File: civitasRuby/pkg/civitasRuby-0.0.1/lib/civitas/sorpresa_salir_carcel.rb
#encoding:utf-8
=begin
Authors: Esther García Gallego
         Yesenia González Dávila
         Grupo B3
=end

module Civitas

class SorpresaSalirCarcel < Sorpresa
  
  def initialize(mazo)
    super("*Salvoconducto*", nil, -1, "", mazo)
  end
  
  def aplicarAJugador (actual, todos)
    if jugadorCorrecto(actual, todos)
      informe(actual, todos)

      tienen_salvoconducto = false
      i = 0
      while i < todos.length && !tienen_salvoconducto
        tienen_salvoconducto = todos[i].tieneSalvoconducto
        i += 1
      end

      if !tienen_salvoconducto
        todos[actual].obtenerSalvoconducto(self)
        salirDelMazo
      end

    end
  end
  
  private # ------------------------------------------------------------------ #
  
  def salirDelMazo
        @mazo.inhabilitarCartaEspecial(self)
    end

  def usada
      @mazo.habilitarCartaEspecial(self)
  end
  
  
end

end

File: P5/civitasJava/src/civitas/SorpresaSalirCarcel.java

package civitas;

import java.util.ArrayList;

public class SorpresaSalirCarcel extends Sorpresa {
    
    SorpresaSalirCarcel (MazoSorpresas mazo) {
        super ("* Salvoconducto *");
        this.mazo = mazo;
        texto = " ¡Obtienes salvoconducto!";
    }
    
    /** Se pregunta a todos los jugadores su alguien tiene la sorpresa para evitar 
     *  la carcel (salvoconducto), si nadie la tiene,la obtiene el jugador actual 
     *  y se llama al método salirDelMazo. */
    @Override
    void aplicarAJugador (int actual, ArrayList<Jugador> todos) {
        if (jugadorCorrecto (actual, todos)) {

            // Se comprueba que el resto de jugadores no tiene el salvoconducto
            boolean tienen_salvoconducto = false;
            for (int i=0 ; i<todos.size() && !tienen_salvoconducto; i++)
                tienen_salvoconducto = todos.get(i).tieneSalvoconducto();

            if (!tienen_salvoconducto) {
                informe (actual, todos);
                todos.get(actual).obtenerSalvoconducto(this);
                salirDelMazo ();
            }
        }
    }
    
    /**  Si la sorpresa es SALIRCARCEL, inhabilita la carta en el mazo. */
    void salirDelMazo () {
        mazo.inhabilitarCartaEspecial(this);
    }

    /** Si la sorpresa es SALIRCARCEL, habilita la carta en el mazo. */
    void usada() {
        mazo.habilitarCartaEspecial(this);
    }
    
}
